Documentation/guides/internals/debug-jni-objrefs.md

```diff
@@ -0,0 +1,414 @@
+
+  * [Via an unhandled exception](#crash-via-unhandled-exception), or
+  * [Via an Android Runtime abort](#crash-dalvik-abort).
+
+In order to diagnose and eventually fix the crash, you will next need to:
+
+ 1. [Collect JNI Object Reference logs](#collect-logs), then
+ 2. [Understand the collected logs](#understand-logs).
+ 2. [Interpreting the collected logs](#read-logs).
+
+You may find [`logcat-parse`](#logcat-parse) useful in reading the log files.
+
+
+<a name="crash-unhandled-exception"></a>
+
+### Crash via Unhandled Exception
+
+When debugging an app crashing via an Unhandled Exception, the debugger
+will show that a `NotSupportedException` was thrown containing an
+inner exception which is a `MissingMethodException`:
+
+![Unhandled Exception dialog with MissingMethodException](debug-jni-objrefs-images/debugger-missingmethodexception.png)
+
+`adb logcat` output will contain:
+
+	F mono-rt : [ERROR] FATAL UNHANDLED EXCEPTION: System.NotSupportedException: Unable to activate instance of type Java.Lang.Runnable from native handle 0x7ff1f3b468 (key_handle 0x466b26f).
+	F mono-rt :  ---> System.MissingMethodException: No constructor found for Java.Lang.Runnable::.ctor(System.IntPtr, Android.Runtime.JniHandleOwnership)
+	F mono-rt :  ---> Java.Interop.JavaLocationException: Exception of type 'Java.Interop.JavaLocationException' was thrown.
+	F mono-rt : Java.Lang.Error: Exception of type 'Java.Lang.Error' was thrown.
+	F mono-rt : 
+	F mono-rt :   --- End of managed Java.Lang.Error stack trace ---
+	F mono-rt : java.lang.Error: Java callstack:
+	F mono-rt :      …
+	F mono-rt :    --- End of inner exception stack trace ---
+	F mono-rt :    at Java.Interop.TypeManager.CreateProxy(Type type, IntPtr handle, JniHandleOwnership transfer) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 348
+	F mono-rt :    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 312
+	F mono-rt :    --- End of inner exception stack trace ---
+	F mono-rt :    at Java.Interop.TypeManager.CreateInstance(IntPtr handle, JniHandleOwnership transfer, Type targetType) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Interop/TypeManager.cs:line 319
+	F mono-rt :    at Java.Lang.Object.GetObject(IntPtr handle, JniHandleOwnership transfer, Type type) in /Users/runner/work/1/s/xamarin-android/src/Mono.Android/Java.Lang/Object.cs:line 304
+	F mono-rt :    …
+
+In particular, `adb logcat` has a `MissingMethodException` mentioning that a constructor
+with the signature `(IntPtr, JniHandleOwnership)` could not be found.
+
+If you see a `NotSupportedException` with
+[*anything else* as the inner exception](https://github.com/xamarin/xamarin-android/issues/7324),
+then this is *not* due to JNI object references, and the rest of this guide will not help you.
```

Comment:
A suggestion to some "next step(s)" if this guide does not help might be useful, even if it says something like "The exception message should tell you what the real error is".